internal/armada/server/authorization.go

```diff
@@ -0,0 +1,158 @@
+package server
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/G-Research/armada/internal/armada/permissions"
+	"github.com/G-Research/armada/internal/armada/repository"
+	"github.com/G-Research/armada/internal/common/auth/authorization"
+	"github.com/G-Research/armada/internal/common/auth/permission"
+	"github.com/G-Research/armada/pkg/api"
+)
+
+// ErrNoPermission represents an error that occurs when a client tries to perform some action
+// through the gRPC API for which it does not have permissions.
+// Produces error messages of the form
+// "Tom" does not own the queue and have SubmitJobs permissions, "Tom" does not have SubmitAnyJobs permissions
+//
+// The caller of a function that may produce this error should capture is using errors.As and prepend
+// whatever action the principal was attempting.
+type ErrNoPermission struct {
+	// Principal that attempted the action
+	Principal authorization.Principal
+	// Reasons that the principal was not allowed to perform the action
+	// For example ["does not own the queue and have SubmitJobs permissions", "does not have SubmitAnyJobs permissions"]
+	Reasons []string
+}
+
+func (err *ErrNoPermission) Error() string {
+	principalName := err.Principal.GetName()
+	reasons := make([]string, len(err.Reasons), len(err.Reasons))
+	for i, reason := range err.Reasons {
+		reasons[i] = fmt.Sprintf("%q %s", principalName, reason)
+	}
+	return strings.Join(reasons, ", ")
+}
+
+// checkReprioritizePerms checks if the principal embedded in the context has permission
+// to reprioritize all of the given jobs. This may require permissions for several queues.
+// If the principal has sufficient permissions, nil is returned. Otherwise an error is returned.
+func (server *SubmitServer) checkReprioritizePerms(ctx context.Context, jobs []*api.Job) error {
+
+	// Set of queues that has at least 1 of the given jobs
```

Comment:
Comment not a tiny bit gratuitous? I'd get a second opinion but comments should be more about why you're doing something rather than what. The code itself is fairly clear

Reply:
I agree. I've updated the comment to say `If the principal has insufficient permissions, an error of type ErrNoPermission is returned.`.